greentasks/scheduler.py

```diff
@@ -1,94 +1,154 @@
-import collections
 import logging
-import uuid
 
-import gevent
+from gevent import spawn_later
+from gevent.queue import Queue, Empty as QueueEmpty
+
+from .tasks import PackagedTask, Task
 
 
 class TaskScheduler(object):
-    QUEUED = 'QUEUED'
-    PROCESSING = 'PROCESSING'
-    NOT_FOUND = 'NOT_FOUND'
+    """
+    A very simple task scheduler built on top of gevent.
+
+    Allows scheduling of periodic, one-off delayed and one-off ordered tasks.
+
+    Periodic tasks will run repeatedly after their execution in the specified
+    amount of time.
+
+    One-off delayed tasks are executed only once after the specified delay has
+    passed.
+
+    One-off ordered tasks are put into a queue and are executed in the same
+    order they were scheduled. The queue is processed in fixed intervals which
+    is specified by the ``consume_tasks_delay`` parameter in the constructor of
+    the scheduler. In each wakeup, exactly one task will be processed.
+    """
+    #: Wrapper class enclosing all the data needed for a single task
+    packaged_task_class = PackagedTask
+
+    #: Base class used to implement the actual task logic
+    base_task_class = Task
 
     def __init__(self, consume_tasks_delay=1):
-        self._queue = collections.OrderedDict()
-        self.current_task = None
+        self._queue = Queue()
         self._consume_tasks_delay = consume_tasks_delay
         self._async(self._consume_tasks_delay, self._consume)
 
-    def _generate_task_id(self):
-        return uuid.uuid4().hex
+    def _async(self, delay, fn, *args, **kwargs):
+        """
+        Schedule a function with the passed in parameters to be executed
+        asynchronously by gevent.
+        """
+        return spawn_later(delay, fn, *args, **kwargs)
 
-    def get_status(self, task_id):
-        if task_id in self._queue:
-            return self.QUEUED
-        if task_id == self.current_task:
-            return self.PROCESSING
-        return self.NOT_FOUND
+    def _execute(self, packaged_task):
+        """
+        Delegate execution of ``packaged_task`` to py:meth:`~PackageTask.run`,
+        which handles the invocation of appropriate callbacks and errbacks and
+        the resolving of the future object.
 
-    def _async(self, delay, fn, *args, **kwargs):
-        return gevent.spawn_later(delay, fn, *args, **kwargs)
+        The return value is forwarded as is, which is the task instance itself,
+        unless instantiation failed for some reason.
+        """
+        return packaged_task.run()
 
-    def _execute(self, task_id, fn, args, kwargs):
-        self.current_task = task_id
+    def _periodic(self, packaged_task):
+        """
+        Execute a periodic task through py:meth:`~TaskScheduler._execute` and
+        reschedule it automatically for the specified ``delay``.
```

Comment:
Thanks, fixed

On Mon, Jun 13, 2016 at 6:47 PM, Manish P Mathai <notifications@github.com>
wrote:

> In greentasks/scheduler.py
> <https://github.com/Outernet-Project/greentasks/pull/1#discussion_r66826394>
> :
>
> >
> > -    def _execute(self, task_id, fn, args, kwargs):
> > -        self.current_task = task_id
> > +    def _periodic(self, packaged_task):
> > +        """
> > +        Execute a periodic task through py:meth:`~TaskScheduler._execute` and
> > +        reschedule it automatically for the specified ``delay``.
>
> delay is not a specified parameter
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Outernet-Project/greentasks/pull/1/files/66f6c9e0d2f4bebd57891c28e8ce66b05df2877a#r66826394>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABRA35IHlBUvFBMT42wKECrRwkTn-9Kfks5qLYmZgaJpZM4I0gRH>
> .
>